### ISSUE TITLE
[Question] Sequential features support guide

### ISSUE THREAD

**Issue by Sharan123** [reczoo/FuxiCTR#120]
Hello to the FuxiCTR team, first of all amazing work with making and keeping this library up-to-date!

I have a question regarding the sequential features support and potentially adding our own sequential feature embeddings. Namely I am aware that the `demo/sample6_DNN_with_sequence_features.py` exist, however the `dataset_config.yaml` does not contain preprocessing from csv, instead we take the data directly from `tiny_seq`. Nowhere do we explicitly specify the sequential features for `click_sequence` outside of the `model_config.yaml` which is connected directly to `DIN` model. 

My question related to this: _**Is there a general example on how we can specify sequential features for any model directly, using the dataset_config.yaml to preprocess from a csv file? How should the feature column in the csv dataset look like in this case**_?

Regarding the part of adding our own sequential feature embeddings, I have found that `fit_sequence_col` in `fuxictr.preprocess.feature_preprocessor.py` sets a default feature_encoder to the feature_map for sequential features - specifically `layers.MaskedAveragePooling()` located at `fuxictr.pytorch.layers.pooling.py`. Looking at the code for various embeddings, we have 

```python
class MaskedAveragePooling(nn.Module):
 # rest of the code ... 
   def forward(self, embedding_matrix, mask=None)
      # with mask
class MaskedSumPooling(nn.Module):
# rest of the code ....
   def forward(self, embedding_matrix)
       # no mask
class KMaxPooling(nn.Module):
# rest of the code ...
   def forward(self, X)
      # with input X
```

There are two questions related to these pooling layers: 
 - _**Are embedding_matrix and matrix X one and the same in the above pooling classes**_, as they are being used in `FeatureEmbeddingDict.forward` in the same way? If so **_what is the shape of this embedding_matrix_**? 
 - **_How could we make our own Embedders for sequential data, for example an LSTMEncoder_** instead of the pooling ops such as the code below
 
```python
 class LSTMEncoder(nn.Module):
    def __init__(self, input_size, hidden_size, num_layers=1, bidirectional=False):
        super(LSTMEncoder, self).__init__()
        self.lstm = nn.LSTM(input_size, hidden_size, num_layers, batch_first=True, bidirectional=bidirectional)

    def forward(self, x):
        outputs, _ = self.lstm(x)
        return outputs
```

Obviously the above code snippet wouldn't work, but how could one make it work as a sequential embedder? (we might also need a pooling layer following the LSTM layer as well, but I am using this as an example)

The last question is regarding shared embeddings between the sequential and categorical data. For example click_sequence and adgroup_id - is my understanding correct that the values in the click sequence are the same categories as ones in adgroup_id in this example and that is why we can share embeddings? Or how do shared embeddings work and when can we used them?

I apologize for the many questions, but I do appreciate your help in this matter!

**Comment by zhujiem** [comment]
Hi! It is a good question. Due to the limit of time, the api doc is not clear enough.

For how to use sequence features, please refer to our SeqCTR benchmark here: https://openbenchmark.github.io/BARS/CTR/leaderboard/amazonelectronics_x1.html

We have benchmarked sequential models on four datasets with sequence features. Take DIN as an example, you can check the README on how to configure and reproduce the experiment. https://github.com/reczoo/BARS/tree/main/ranking/ctr/DIN/DIN_amazonelectronics_x1

Generally, to implement your own pooling, I suggest to follow the DIN model implmentation. When you configure `feature_encoder: null`, you will obtain the sequence feature in the forward method.

```
self.lstm = LSTM()
self.pooling = layers.MaskedAveragePooling()

def forward(self, inputs):
    X = self.get_inputs(inputs)
    feature_emb_dict = self.embedding_layer(X)
    seq_feature_embs = feature_emb_dict["SEQ_FEATURE_NAME"] # (NONE, seq_len, emb_size)
    lstm_h = self.lstm(seq_feature_embs )
    pool_out = self.pooling(lstm_h)

```

For share_embedding, if different features share the same vocabulary, for example, seq_id and target_id or seq_category and target_category, you could choose to share a same embedding table. It serves as a hyper-parameter for model tuning. You could find the detailed example in the above [DIN_amazonelectronics_x1](https://github.com/reczoo/BARS/tree/main/ranking/ctr/DIN/DIN_amazonelectronics_x1). 

Note that `feature_specs: xxx` can be set to update the key-value pairs in feature_cols.

**Comment by Sharan123** [comment]
Thank you for the answer! It is much clearer - I was thinking of using sequential features for click prediction which is slightly different than the Amazon Electronics x1 Dataset task. Would you still recommend I use the targets for the specific sequence to train the 'sequence embedding part' of the network separately (like in DIN for amazonelectronics) or should I only use the embedding and backprop through the entire network for the BCE loss, and let the model learn implicit embeddings for sequential features that way? 

So to confirm about the `feature_encoder: null`, because the load method loads first the model then the dataset, we would overwrite the `feature_encoder: null` with the ones in the dataset config - `feature_encoder: layers.MaskedAveragePooling()`
which then in [fuxictr/preprocess/feature_preprocessor.py](https://github.com/reczoo/FuxiCTR/blob/b026ce948108b7cca926be4016c62993ee5740c7/fuxictr/pytorch/layers/embeddings/feature_embedding.py#L79) would be assigned in the 
```python
if feature_spec.get("feature_encoder", None):
      self.feature_encoders[feature] = self.get_feature_encoder(feature_spec["feature_encoder"])
```
where `self.feature_encoders[feature]` is being called in `FeatureEmbeddingDict.forward` method. Is this understanding correct? 

In this case why do we still need to specify them in `model_config.yaml`, wouldn't the `dataset_config.yaml` specification for the feature_encoder suffice?

**Comment by zhujiem** [comment]
1. I think your first question is an open problem. I guess you can try both and compare them.

2. If you set `feature_encoder: null`, the sequence will not be pooled. If the key `feature_encoder` is absent, the default layers.MaskedAveragePooling() will be used.
